### PR TITLE
use clang 12

### PR DESCRIPTION
**Pull request by lizan**
Signed-off-by: Lizan Zhou <lizan@tetrate.io>

### CODE REVIEW

**Comment by moderation** [comment]
Can we bump the Clang version at https://github.com/envoyproxy/envoy/blob/main/bazel/repository_locations.bzl#L902-L916 for consistency?

**Comment by moderation** [comment]
It would be preferable to not introduce a patch. Since we control `envoy-build-tools` can this not be fixed upstream?

**Comment by lizan** [comment]
@moderation this is just a draft to test how this goes in CI, will see what will be the path forward.

**Comment by yanavlasov** [comment]
Coverage dip maybe due to tooling change? We can adjust the levels for now. @alyssawilk

**Comment by lizan** [comment]
Yes I think coverage dip is due to they change how lines get counted. I will dig it later.

**Comment by alyssawilk** [comment]
If we need to bump we need to bump.  The problem is that it looks like it's now considering comments to be code.  It'd be lovely to fix if we can

see 115 for 
https://storage.googleapis.com/envoy-pr/cb08fb4/coverage/source/common/api/posix/os_sys_calls_impl.cc.gcov.html
vs
https://storage.googleapis.com/envoy-postsubmit/main/coverage/source/common/api/posix/os_sys_calls_impl.cc.gcov.html

**Comment by lizan** [comment]
> If we need to bump we need to bump. The problem is that it looks like it's now considering comments to be code. It'd be lovely to fix if we can
> 
> see 115 for
> https://storage.googleapis.com/envoy-pr/cb08fb4/coverage/source/common/api/posix/os_sys_calls_impl.cc.gcov.html
> vs
> https://storage.googleapis.com/envoy-postsubmit/main/coverage/source/common/api/posix/os_sys_calls_impl.cc.gcov.html

This is an improvement? The main branch shows comments to be code and PR doesn't count that.

**Comment by alyssawilk** [comment]
yep, was reading it backwards.
As far as I can tell, old coverage considered comments code, new coverage does not, and the ratio of covered comments to non-covered comments makes a difference in aggregate :-P
I'm fine bumping the threshhold but if we can please only bump it for the project as a whole (but keep the bar for each directory at the 96.6) I think it'll be easier to gain that ground back.

**Review by yanavlasov — APPROVED** [review]

**Comment by lizan** [comment]
@htuch @moderation this is ready to go, about patches:

rules_cc patch posted to https://github.com/bazelbuild/rules_cc/pull/115
envoy-build-tools patch, while we control the repo, though this is patching generated toolchain config which is coming from bazel itself and it is not easy to override how it is generated. If we move this to envoy-build tools it will be a `patch -p1` in the generator script and isn't any better.

**Comment by yanavlasov** [comment]
> @htuch @moderation this is ready to go, about patches:
> 
> rules_cc patch posted to [bazelbuild/rules_cc#115](https://github.com/bazelbuild/rules_cc/pull/115) envoy-build-tools patch, while we control the repo, though this is patching generated toolchain config which is coming from bazel itself and it is not easy to override how it is generated. If we move this to envoy-build tools it will be a `patch -p1` in the generator script and isn't any better.

I can work with the RBE dev to make this aspect of toolchain generator configurable. Once this is done we can update toolchain generator. I see this patch as temporary.

**Comment by alyssawilk** [comment]
hey, did we envoy-dev announce this change?  we should warn folks to install clang 12 no?  Lizan if this makes sense can you do that?

**Comment by wrowe** [comment]
AIUI, the build tools are bumped to 12. The flags we've changed should keep clang-11 users largely compatible, there's nothing revolutionary in this change, so they should still be able to build locally with clang-11 except very specific platforms.

File: api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto [hunk 1]
```diff
@@ -390,7 +390,8 @@ message FilterStateRule {
 
   // A map of string keys to requirements. The string key is the string value
   // in the FilterState with the name specified in the *name* field above.
-  map<string, JwtRequirement> requires = 3;
+  map<string, JwtRequirement>
+  requires = 3;
```
Comment (sergiitk) [inline api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:394]: I wonder if this change was accidental. `FORCE_PROTO_FORMAT=yes ./tools/proto_format/proto_format.sh check` complains about this. 
(same applies to corresponding v3/config.proto)

**Pull request merged**